### PR TITLE
Aggregate picard duplication metrics from multiple libraries

### PR DESCRIPTION
**Pull request by mlaszloffy**

### CODE REVIEW

File: bam_qc_metrics/bam_qc.py [hunk 1]
```diff
@@ -273,56 +288,156 @@ def read_mark_dup(self, input_path):
...
+
+            # find value of M, large enough to act as other side for bisection method
+            while f(M * unique_read_pairs, unique_read_pairs, read_pairs) > 0:
+                M *= 10.0
+
+            # use bisection method (no more than 40 times) to find solution
+            for i in range(40):
+                r = (m + M) / 2.0;
+                u = f(r * unique_read_pairs, unique_read_pairs, read_pairs);
+                if u == 0:
```
Comment (iainrb) [inline bam_qc_metrics/bam_qc.py]: `u` is unlikely to be exactly zero, within floating point accuracy.
We could have a constant delta (set to a small positive value, eg. 0.000001) and have:

```
if math.fabs(u) < delta:
    break
elif u > 0:
    m = r
elif u < 0:
    M = r
```

File: bam_qc_metrics/bam_qc.py [hunk 2]
```diff
@@ -273,56 +288,156 @@ def read_mark_dup(self, input_path):
...
                 section += 1
+            elif section == 2:
+                # DuplicationMetrics record
+                library_values = re.split("\t", line)
+                values.append(library_values)
             elif section == 3 and re.match('## HISTOGRAM', line):
+                # histogram start
                 section += 1
-            elif section == 4 and re.match("BIN\tVALUE", line):
+            elif section == 4 and re.match("BIN\tCoverageMult", line):
```
[the remark this answers is not in the document]
Comment (mlaszloffy) [reply bam_qc_metrics/bam_qc.py]: Fixed - this change removed support for picard1 histogram.

File: bam_qc_metrics/bam_qc.py (hunk 1 above)
Comment (mlaszloffy) [reply bam_qc_metrics/bam_qc.py]: Good call, added check against machine epsilon

File: bam_qc_metrics/bam_qc.py [hunk 3]
```diff
@@ -273,56 +288,156 @@ def read_mark_dup(self, input_path):
...
+                value.append(0)
+            elif len(keys) != len(value):
+                # otherwise, mismatched key/value totals are an error
+                msg = "Numbers of keys and values in %s do not match" % input_path
+                self.logger.error(msg)
+                raise ValueError(msg)
+
+            # parse
+            metrics = {}
+            for i in range(len(keys)):
```
[the remark this answers is not in the document]
Comment (mlaszloffy) [reply bam_qc_metrics/bam_qc.py]: Yes this is cleaner - fixed.

**Comment by mlaszloffy** [comment]
I was planning on exporting only the histograms that are produced by picard.  For example, when there are multiple libraries as input to MarkDuplicates, the `CoverageMult` (aka ROI) histogram is not produced.  Exporting an empty `CoverageMulti` histogram may introduce some confusion?

**Pull request merged**